### ISSUE TITLE
Boot cmdline arguments to set language, keyboard, etc.

### ISSUE THREAD

**Issue by probonopd** [SUSE/kiwi#751]
## Problem description

I would like to be set the language and the keyboard language for all Live ISOs produced by kiwi-ng by specifying boot cmdline arguments, and without needing to manually change the contents of the live ISOs. Other things that would be nice to be able to configure using boot cmdline arguments are the timezone, whether UTC is used, the user name, and the hostname of the machine.

Other Live ISO systems have this functionality:

* __Ubuntu__ uses _casper_ which has `console-setup/layoutcode=` and `locale=` ([Docs](http://manpages.ubuntu.com/manpages/artful/man7/casper.7.html))
* __Debian__ uses _live-config_ which has `live-config.keyboard-layouts=` and `live-config.locales=` ([Docs](https://manpages.debian.org/jessie/live-config-doc/live-config.7.en.html))
* __Mageia__ uses _draklive2_ which has `lang=` ([Docs](https://wiki.mageia.org/en/Draklive2#Language_selection_menu))
* __Fedora/CentOS__ use _dracut_ which has `vconsole.keymap=` and `locale.LANG=` ([Docs](http://man7.org/linux/man-pages/man7/dracut.cmdline.7.html))

and so on.

This would be especially useful for my use case where I boot different Live ISOs of different distributions using something like [SystemImageKit](https://github.com/probonopd/SystemImageKit), but would still like all the distributions use the correct language and keyboard language.

## Expected behaviour

Boot the Live ISO from GRUB with something along the lines of

```
isofrom_device=/dev/sda1 isofrom_system=/boot/iso/...iso keytable=de language=de_DE.UTF-8 ramdisk_size=512000 ramdisk_blocksize=4096 splash=silent quiet showopts
```

Note the `keytable=de language=de_DE.UTF-8` (this is just an example - the actual names of the variables could be different as there seems to be no real cross-distribution standard for this yet)

Following the discussion in https://github.com/openSUSE/kiwi/issues/355 I was using [this hack](https://github.com/probonopd/SystemImageKit/blob/34af5cb8a13d6e8d5816ebc6cd6008d08bbf407f/boot/iso/additional-initramfs/kiwi-international/kiwi-hooks/postmount.sh) but it seems not to work anymore, and also I would like to see this functionality to be integrated into kiwi-ng properly.

cc @Vogtinator

**Comment by Vogtinator** [comment]
The openSUSE live media uses live-langset, which has the same "lang=" parameter as Mageia does.

**Comment by probonopd** [comment]
Thanks @Vogtinator. Following questions:
* Is there a central place where all the command line options are documented, something like http://man7.org/linux/man-pages/man7/dracut.cmdline.7.html?
* Can we expect this to work on all variants of *SUSE generated by kiwi-ng?

**Comment by Vogtinator** [comment]
The live-langset-data package takes care of that, if it's installed the lang parameter will work, if not it won't.

All openSUSE live media uses it.

> Is there a central place where all the command line options are documented, something like http://man7.org/linux/man-pages/man7/dracut.cmdline.7.html?

Yes, there ^. It uses dracut for booting, so that still applies. There are some parameters interpreted by the kiwi dracut modules.

> How about setting the keyboard language, and the other settings mentioned above?

live-langset-data takes care of that.

That actually reminds me - isolinux sets both lang and keytable, but keytable isn't actually implemented in live-langset yet, it always uses the one defined by lang. I'll fix that.

**Comment by davidcassany** [comment]
@probonopd I am not fully aware of which boot options could be used to configure keytable, language, timezone and so on. However, even I know it is not exactly the same, in Kiwi one can configure some things using the `<preferences>` section of the description file. You can set keytable, locale and timezone. See [preferences](http://suse.github.io/kiwi/development/schema.html#id36) tag docs.

**Comment by probonopd** [comment]
> in Kiwi one can configure some things using the <preferences> section of the description file

Yes, I am aware of that, but these are _build time_ options whereas I am looking for _boot time_ options.

>  It uses dracut for booting, so that still applies

Let me repeat that so that I get this right: We can use all the boot-time options known from dracut, because kiwi-ng uses dracut? Now that'd be cool :-)

**Comment by Vogtinator** [comment]
> Let me repeat that so that I get this right: We can use all the boot-time options known from dracut, because kiwi-ng uses dracut? Now that'd be cool :-)

Yes! It's opt-in though, depending on the image description.

**Comment by Vogtinator** [comment]
[OT]

> Current Kiwi uses dracut as the only valid initrd system for vmx and iso images.

Did that get changed recently? I had to add `<package name="dracut-kiwi-live"/>` manually to enable it.

[/OT]

**Comment by probonopd** [comment]
Is this whole "how to use the Live ISOs" thing documented somewhere, e.g., in the Wiki? I did find a couple of "loose" forum entries and such but not an official-looking documentation. I'd like to help with documentation if I can.

**Comment by davidcassany** [comment]
@Vogtinator it is not that including` <package name="dracut-kiwi-live"/>` enables it. You must include this package in any ISO that makes use of the dracut-kiwi-live module. In fact you could avoid it using the `dmsquash` module (`flags="dmsquash"`) once #749 is merged.

**Comment by probonopd** [comment]
fwiw, I think it would be good if _all_ Live images created by kiwi-ng behaved as _consistently_ as possible with regard to their "Live system features". It is cumbersome for users to figure out how stuff works in every different Live ISO because there are build-time options that affect the functionality of the generated Live ISOs. For example, all Ubuntu Live ISOs have been behaving pretty _consistent_ over the years concerning the "Live system features", whereas the *SUSE ones seemed to be less "predictable".

**Comment by probonopd** [comment]
Can you please give a concrete example on how to set the keyboard to German. With this command line it does __not__ work for me (in the booted system, the keyboard is still English):

linux@localhost:~> cat /proc/cmdline 
BOOT_IMAGE=(loop)/boot/x86_64/loader/linux iso-scan/filename=/boot/iso/openSUSE-Tumbleweed-Rescue-CD-x86_64-Current.iso selinux=0 rd.live.user=me rd.live.host=host __vconsole.keymap=de locale.LANG=en_US.UTF-8__ max_loop=256 initrd=initrd splash=silent quiet root=live:CDLABEL=openSUSE_Tumbleweed_Rescue_CD rd.live.image rd.live.overlay.persistent rd.live.overlay.cowfs=ext4 showopts

**Comment by Vogtinator** [comment]
`LANG=de` or `LANG=de_DE`.

**Comment by probonopd** [comment]
Does not seem to work for me:

linux@localhost:~> cat /proc/cmdline 
BOOT_IMAGE=(loop)/boot/x86_64/loader/linux iso-scan/filename=/boot/iso/openSUSE-Tumbleweed-Rescue-CD-x86_64-Current.iso selinux=0 rd.live.user=me rd.live.host=host max_loop=256 __LANG=de__ initrd=initrd splash=silent quiet root=live:CDLABEL=openSUSE_Tumbleweed_Rescue_CD rd.live.image rd.live.overlay.persistent rd.live.overlay.cowfs=ext4 showopts

Keyboard is still English.

**Comment by Vogtinator** [comment]
Sorry, it's actually lowercase: `lang=de` or `lang=de_DE`.

**Comment by probonopd** [comment]
Unfortunately not working either, at least on openSUSE-Tumbleweed-Rescue-CD-x86_64-Current.iso

**Comment by Vogtinator** [comment]
Indeed - `lang=de` is broken while `lang=de_DE` works fine. isolinux always uses the full code with country so that shouldn't matter.

**Comment by probonopd** [comment]
Thank you very much @Vogtinator, `lang=de_DE` works indeed. But now both the system language and the keyboard language are in German, whereas I would like the system language to remain English. Is that possible?

**Comment by schaefi** [comment]
After reading through this thread it seems there is no defect in kiwi itself which we need to care for.

**Comment by Vogtinator** [comment]
The fix is already done, I just need to submit it.

If you really want to, you can open a report on bugzila, this is unrelated to kiwi.

**Comment by probonopd** [comment]
Thanks @Vogtinator if you are already about to submit it, then I don't need to open an issue for it. But can you point me to the correct project on Bugzilla where I should submit similar issues in the future, if here is not the correct place? Thanks.

**Comment by probonopd** [comment]
Hi @Vogtinator was the fix submitted, can you point me to the commit? Thanks.

**Comment by Vogtinator** [comment]
@probonopd: Yes, the latest TW lives should support keytable= now: https://build.opensuse.org/package/rdiff/openSUSE:Factory/live-langset-data?linkrev=base&rev=7

**Comment by probonopd** [comment]
`iso-scan/filename=/boot/iso/openSUSE-Tumbleweed-KDE-Live-x86_64-Snapshot20180606-Media.iso (...) keytable=de_DE` does __not__ seem to work for me.

**Comment by Vogtinator** [comment]
de_DE is not a keyboard layout - that's a country code. `keytable=de` should work.

**Comment by probonopd** [comment]
Where does one know this stuff from? Is there a wiki page or manpage for it? Thanks.

**Comment by Vogtinator** [comment]
In this case it should be fairly obvious - keytable sets the keyboard layout so the list of valid layouts can be viewed with `localectl list-keymaps`.

**Comment by probonopd** [comment]
On `openSUSE-Tumbleweed-KDE-Live-x86_64-Snapshot20180606-Media.iso`:

```
linux@host:~> localectl list-keymaps
Failed to read list of keymaps: No such file or directory
```

**Comment by Vogtinator** [comment]
I can confirm that - it's actually a bug in the kernel!

The live cd uses overlayfs to provide a rw root on top of a ro medium, which causes files to have a device number other than the directory they're inside of:

```
localhost:~ # stat /usr/share/kbd/keymaps/xkb/
  File: /usr/share/kbd/keymaps/xkb/
  Size: 20480           Blocks: 40         IO Block: 4096   directory
Device: 26h/38d Inode: 22505       Links: 2
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2018-05-10 22:27:11.591178363 +0200
Modify: 2018-05-10 22:21:27.547169278 +0200
Change: 2018-05-10 22:27:11.591178363 +0200
 Birth: -
localhost:~ # stat /usr/share/kbd/keymaps/xkb/de.map.gz 
  File: /usr/share/kbd/keymaps/xkb/de.map.gz
  Size: 3850            Blocks: 8          IO Block: 4096   regular file
Device: 27h/39d Inode: 50521       Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2018-05-10 22:27:11.351178357 +0200
Modify: 2018-04-16 14:24:07.000000000 +0200
Change: 2018-05-10 22:27:11.351178357 +0200
 Birth: -
```

localectl uses nftw with FTW_MOUNT to traverse the tree of /usr/share/keymaps, but never gets to see any of the files within as they are not part of the same device...

Can you file a bug report?

**Comment by Vogtinator** [comment]
See also https://patchwork.kernel.org/patch/9427507/ - I don't think that patch is correct though.

**Comment by probonopd** [comment]
> Can you file a bug report?
Where?

**Comment by Vogtinator** [comment]
bugzilla.opensuse.org - the kernel maintainers should handle this then.

**Comment by probonopd** [comment]
Submitted at https://bugzilla.opensuse.org/show_bug.cgi?id=1096807
